Longhorn Manual Resolution (#1963)
spectrocloud/librarium: docs/docs-content/integrations/longhorn.md

## Version Supported

<Tabs queryString="versions">
<TabItem label="1.5.x" value="1.5.x">

## Prerequisites

- Kubernetes cluster is 1.21 or higher.

## Parameters

The table lists commonly used parameters you can configure when adding this pack.

| Parameter                | Description                                            | Default                                     |
|-------------------------|--------------------------------------------------------|---------------------------------------------|
| `namespace`               | The namespace where the Longhorn will be installed. | `longhorn-system` |
| `defaultClass`            | The volume type to be used. | `true` |
| `defaultFsType`           | The default file system.    | `ext4` |
| `defaultClassReplicaCount`| The default number of copies of data store in your cluster.   | `3`         |

> | `defaultDataLocality`     | The default location where data computation will occur. | `disabled` Best effort |

Reviewer: This is a little confusing to me. If it's a location, why is the default "disabled"? And maybe we can expand on the "best effort" a bit?

Author: Agree, I'm gonna also link to their parameter page.

> | `reclaimPolicy`           | This means that a dynamically provisioned volume will be automatically deleted when deletes when corresponding PersistentVolumeClaim is deleted. For important data, it is more appropriate to use the "Retain" policy | `Delete`  |

Reviewer: ```suggestion
| `reclaimPolicy`           | The policy that regulates dynamically provisioned volumes when its corresponding PersistentVolumeClaim is deleted. The default policy is `Delete`. This means that a dynamically provisioned volume will be automatically deleted when corresponding PersistentVolumeClaim is deleted. For important data, it is more appropriate to use the "Retain" policy. | `Delete`  |
```

> | `migratable`              | The ability to transfer data to another data storage systems | `false`   |

Reviewer: ```suggestion
| `migratable`              | The ability to transfer data to another data storage systems. | `false`   |
```

| `recurringJobSelector.enable`   | The management of recurring jobs. You can enable this feature and type a comma-separated list of jobs to run: `recurringJobSelector.enable.jobList [ ]`  | `false`   |

## Usage

Longhorn provides these features:

- Enterprise-grade distributed storage with no single point of failure.
- Incremental snapshots of block storage.
- Backup to secondary storage (NFSv4 or S3-compatible object storage) built on change block detection.
- Recurring snapshot and backup.

For more information, check out Longhorn guide on [How to Create Volumes](https://longhorn.io/docs/1.4.0/volumes-and-nodes/create-volumes/).

## Troubleshooting

Below are some common issues you may encounter when using Longhorn. Use the following information to help you troubleshoot and resolve the issue.

### Deadlock With RWX Mode

> Starting with version 1.5.X, a known issue exists where a deadlock may occurs when using a RWX volume and a cluster node is recreated. This issue is disclosed as a known issue in the [Longhorn GitHub repository](https://github.com/longhorn/longhorn/wiki/Release-Known-Issues#v153).   

Reviewer: ```suggestion
Starting with version 1.5.X, a known issue exists where a deadlock may occur when using a RWX volume and a cluster node is recreated. This issue is disclosed as a known issue in the [Longhorn GitHub repository](https://github.com/longhorn/longhorn/wiki/Release-Known-Issues#v153).   
```

To resolve the deadlock, you must manually update the status of the Longhorn Storage Manager. Use the following command to resolve the deadlock. Replace `PVC-ID` with the ID of the Persistent Volume Claim (PVC) that is in the `longhorn-system` namespace that and in **RWX** mode.

```bash
kubectl --namespace longhorn-system patch lhsm PVC-ID \
--type=merge --subresource status --patch 'status: {state: error}'
```

:::tip

> To identify the ID of the PVC, use the following command to list all PVCs in the `longhorn-system` namespace. Seach for the PVC that has the **ACCESS MODE** set to **RWX**.

Reviewer: ```suggestion
To identify the ID of the PVC, use the following command to list all PVCs in the `longhorn-system` namespace. Search for the PVC that has the **ACCESS MODE** set to **RWX**.
```

| `charts.longhorn.global.persistence.defaultDataLocality`     | The default location where data computation will occur. | `disabled`|
| `charts.longhorn.global.persistence.reclaimPolicy`           | The policy that regulates dynamically provisioned volumes when its corresponding PersistentVolumeClaim is deleted. The default policy is `Delete`. This means that a dynamically provisioned volume will be automatically deleted when corresponding PersistentVolumeClaim is deleted. For important data, it is more appropriate to use the "Retain" policy. | `Delete`  |
| `charts.longhorn.global.persistence.migratable`              | The ability to transfer data to another data storage systems. | `false`   |
| `charts.longhorn.global.persistence.recurringJobSelector.enable`   | The management of recurring jobs. You can enable this feature and type a comma-separated list of jobs to activate.  | `false`   |

## Usage

The official Longhorn documentation contains several guides and examples. The following guides are a good place to start when learning how to use Longhorn.

> - [How to Create Volumes](https://longhorn.io/docs/1.5.3/volumes-and-nodes/create-volumes/).

Reviewer: ```suggestion
- [How to Create Volumes](https://longhorn.io/docs/1.5.3/volumes-and-nodes/create-volumes/)
```
